lib/Basic/FileSystem.cpp

```diff
@@ -62,6 +62,19 @@ class LocalFileSystem : public FileSystem {
     }
     return std::unique_ptr<llvm::MemoryBuffer>(result->release());
   }
+
+  virtual bool remove(const std::string& path) override {
```

Comment:
I think this is up to the client to manage what items they put in the list. The reason why this is important is that some build steps won't ever be able to accurately enumerate all of the things they "might" put in an output put (e.g., a task which just blindly does a `cp -r` from a user directory).

Also, another thing we should do here is always attempt to remove empty directories up the stale paths... it might be worth allowing this tool to also have a list of "roots" past which it would never traverse for that check.